coreax/kernel.py

```diff
@@ -622,14 +685,246 @@ def divergence_x_grad_y_elementwise(self, x: ArrayLike, y: ArrayLike) -> Array:
+    r"""
+    Define a periodic kernel.
+
+    Given :math:`\lambda =`'length_scale',  :math:`\rho =`'output_scale', and
+    :math:`\p =`'periodicity', the periodic kernel is defined as
+    :math:`k: \mathbb{R}^d\times \mathbb{R}^d \to \mathbb{R}`,
+    :math:`k(x, y) = \rho * \exp(\frac{-2 \sin^2(\pi ||x-y||/p)}{\lambda^2})` where
+    :math:`||\cdot||` is the usual :math:`L_2`-norm.
+
+    .. note::
+        Note that the Periodic kernel is not differentiable when :math:`x=y`.
+
+    :param length_scale: Kernel smoothing/bandwidth parameter, :math:`\lambda`
+    :param output_scale: Kernel normalisation constant, :math:`\rho`
+    :param periodicity: Parameter controlling the periodicity of the kernel. :\math: `p`
+    """
+
+    length_scale: float = 1.0
+    output_scale: float = 1.0
+    periodicity: float = 1.0
+
+    @override
+    def compute_elementwise(self, x: ArrayLike, y: ArrayLike) -> Array:
+        return self.output_scale * (
+            jnp.exp(
+                -2
+                * jnp.sin(
+                    jnp.pi * jnp.linalg.norm(jnp.subtract(x, y)) / self.periodicity
+                )
+                ** 2
+                / self.length_scale**2
+            )
+        )
+
+    @override
+    def grad_x_elementwise(self, x: ArrayLike, y: ArrayLike) -> Array:
+        return -self.grad_y_elementwise(x, y)
+
+    @override
+    def grad_y_elementwise(self, x: ArrayLike, y: ArrayLike) -> Array:
+        dist = jnp.linalg.norm(jnp.subtract(x, y))
+        body = jnp.pi * dist / self.periodicity
+        return (
+            (
+                4
+                * jnp.subtract(x, y)
+                * self.output_scale
+                * jnp.pi
+                / (dist * self.periodicity * self.length_scale**2)
+            )
+            * jnp.sin(body)
+            * jnp.cos(body)
+            * jnp.exp(-(2 / self.length_scale**2) * jnp.sin(body) ** 2)
+        )
+
+    @override
+    def divergence_x_grad_y_elementwise(self, x: ArrayLike, y: ArrayLike) -> Array:
+        d = len(jnp.asarray(x))
+        sub = jnp.subtract(x, y)
+        dist = jnp.linalg.norm(sub)
+        factor = jnp.pi / self.periodicity
+        func_body = factor * dist
+        grad_factor = sub / dist
+        output_factor = 4 * factor * self.output_scale / self.length_scale**2
+
+        func_1 = 1 / dist
+        func_2 = jnp.sin(func_body)
+        func_3 = jnp.cos(func_body)
+        func_4 = jnp.exp(-(2 / self.length_scale**2) * func_2**2)
+
+        first_term = func_1 * func_2 * func_3 * func_4
+        second_term = (
+            -grad_factor * func_1**2 * func_2 * func_3 * func_4
+            - grad_factor * factor * func_1 * func_2**2 * func_4
+            + grad_factor * factor * func_1 * func_3**2 * func_4
+            - (output_factor * sub * func_1**2 * func_2**2 * func_3**2 * func_4)
+        )
+
+        return output_factor * (d * first_term + jnp.dot(second_term, sub))
+
+
+class LocallyPeriodicKernel(ProductKernel):
```

Comment:
In this class, we assume the output scale and length scale are the same for both the inner RBF & periodic kernels. Might it be desired by the user to set these separately?

Reply:
I think the docstring parameter entry for squared exponential output scale needs updating, currently reads:

```
    :param squared_exponential_output_scale: SquaredExponential Kernel
        smoothing/bandwidth parameter, :math:`\lambda`
```

on line 903-904

Reply:
Good spot, fixed